### PR TITLE
Generalize `lookup`

### LINKED ISSUE #2674: Generalizations of `lookup`
[lookup](https://scipp.github.io/generated/functions/scipp.lookup.html?highlight=lookup#scipp.lookup) is used for mapping values of a variable to derived values, based on a "stepwise function" defined based on a histogram. This is used, e.g., to make time-stamps (of individual neutron events) to sample environment log values such as a temperature reading. For more detailed background see also scipp/scippneutron#237.

In practice sample environment logs may have slightly different interpretations (it may not define a "histogram"), e.g., we may want to use the previous or nearest data point. This is similar to options of https://scipp.github.io/generated/modules/scipp.interpolate.interp1d.html#scipp.interpolate.interp1d, but we may need additional options for dealing with noisy fluctuating logs, etc. The related Mantid algorithm [GenerateEventsFilter](https://docs.mantidproject.org/nightly/algorithms/GenerateEventsFilter-v1.html) may give an idea of some of the requirements.

Notes:

- This is not exactly `interp1d` since:
  - We need a very fast implementation since this will need to perform a "lookup" for every event of binned data.
  - We want to use this with `transform_coords`, so a variable containing values should be returned, not a data array.
- `lookup` is not a very clear/user-friendly name. If we envision a set of similar functions, can we come up with a better naming scheme?
- `Lookup` could be extended with `__call__` (doing the same as `__getitem__`) for support of `transform_coords`, but this would still require some manual steps to get the sample environment log into the required histogram layout.

### PR DESCRIPTION
**Pull request by SimonHeybrock**
This got more complex than anticipated during our initial discussions (an indicator that it is probably important not to leave this to the user, since it is relatively easy to get wrong).

- Add `fill_value` argument.
- Change default `fill_value` to NaN (away from 0), if dtype supports it.
- Fix a bug that led to ignored masks, caused by a previously added optimization.
- Support non-histogram "functions", with modes `linear` and `previous`. 
- Add `__call__`, which does the same as `__getitem__`. Should we consider removing `__getitem__`?

Furthermore, when trying to use `sc.lookup` with `transform_coords`, I ran into two problems:

- Objects defining `__call__` include `self` in the argument list, which broke `transform_coords`. This is fixed (in an ugly and not entirely safe way, I did not find to detect this in a 100% safe manner).
- `__name__` is not defined for various callables, e.g., objects created by `functools.partial`. This is also fixed (bypassed).
- `sc.lookup` returns a callable, but to work with `transform_coords` we would need to have a `__call__` that expects a parameter with name matching the dimension of the lookup function. This is tricky. I chose to update `transform_coords` to use a custom property of the callable.

Fixes #2674.

### CODE REVIEW

**Comment by SimonHeybrock** [comment]
Wondering if `mode='previous'` would be a better default? Or should there be no default at all?

File: tests/lookup_test.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+
+@pytest.mark.parametrize("dtype", ['bool', 'int32', 'int64', 'float32', 'float64'])
+def test_nearest(dtype):
+    x = sc.linspace(dim='xx', start=0, stop=1, num=5)
+    data = sc.array(dims=['xx'], values=[0, 1, 0, 2, 2], dtype=dtype)
+    da = sc.DataArray(data=data, coords={'xx': x})
+    var = sc.array(dims=['event'], values=[0.1, 0.4, 0.1, 0.6, 0.9, 1.1, 0.2])
+    expected = sc.array(dims=['event'], values=[0, 0, 0, 0, 2, 2, 1], dtype=dtype)
+    assert sc.identical(sc.lookup(da, mode='nearest')(var), expected)
```
Comment (jl-wynen) [inline tests/lookup_test.py:69]: This and the test above are missing cases for points below 0.

File: lib/core/include/scipp/core/element/event_operations.h [hunk 2]
```diff
@@ -78,27 +78,38 @@ constexpr auto map = overloaded{
...
+                       const auto &weights, const auto &fill) {
+                 auto it = std::upper_bound(edges.begin(), edges.end(), coord);
+                 using T = std::decay_t<decltype(get(weights, 0))>;
+                 return (it == edges.end() || it == edges.begin())
+                            ? fill
+                            : get(weights, --it - edges.begin());
+               }};
+
+constexpr auto lookup_previous =
+    overloaded{map, [](const auto &point, const auto &x, const auto &weights,
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply lib/core/include/scipp/core/element/event_operations.h:107]: `edges` does not make sense in this case?

File: src/scipp/core/bins.py [hunk 3]
```diff
@@ -1,31 +1,54 @@
...
             # Significant speedup if `func` is large but mostly constant.
-            func = merge_equal_adjacent(func)
+            if op == _cpp.buckets.map:
+                if not islinspace(func.coords[dim], dim).value:
+                    func = merge_equal_adjacent(func)
+            else:
+                # In this case the C++ implementation currently used no linspace
+                # optimization, so the extra check is skipped.
+                transition = func.data[:-1] != func.data[1:]
+                func = concat([func[0], func[1:][transition]], dim)
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply src/scipp/core/bins.py:33]: Which part do you think is expensive?

File: src/scipp/coords/rule.py [hunk 4]
```diff
@@ -194,9 +201,17 @@ def rule_output_names(rules: List[Rule], rule_type: type) -> Iterable[str]:
...
 
 
-def _arg_names(func) -> Tuple[str]:
+def _arg_names(func) -> Dict[str, str]:
     spec = inspect.getfullargspec(func)
     if spec.varargs is not None or spec.varkw is not None:
         raise ValueError('Function with variable arguments not allowed in '
                          f'conversion graph: `{func.__name__}`.')
-    return tuple(spec.args + spec.kwonlyargs)
+    if inspect.isfunction(func) or func.__class__ == partial:
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply src/scipp/coords/rule.py:209]: I think lambdas worked even before this change (and I think I added a test)?

File: tests/coords/transform_coords_test.py [hunk 5]
```diff
@@ -1003,3 +1004,35 @@ def test_raises_when_keyword_syntax_clashes_with_options(option):
...
+    class A:
+
+        def __call__(self, x):
+            return x * x
+
+    da = sc.data.table_xyz(nrow=10)
+    assert 'xx' in da.transform_coords(xx=A()).coords
+
+
+def test_input_coords_can_be_defined_via_property():
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply tests/coords/transform_coords_test.py:1030]: But how would this work with `Lookup`?

File: src/scipp/core/bins.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jl-wynen) [reply src/scipp/core/bins.py:33]: Comparing arrays and indexing by mask.

File: src/scipp/coords/rule.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jl-wynen) [reply src/scipp/coords/rule.py:209]: True.

File: tests/coords/transform_coords_test.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jl-wynen) [reply tests/coords/transform_coords_test.py:1030]: I guess, `Lookup` would inherit from `Node`. Not very nice...
I was somehow under the impression that `Lookup` would be able to do something on its end. But it doesn't know that it is being passed to `transform_coords`.

What about a more unique name like `__transform_coords_inputs__` to avoid potential (unlikely) conflicts?

File: src/scipp/core/bins.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply src/scipp/core/bins.py:33]: I had done comparisons when originally implementing the existing branch. Obviously there are corner cases where it is slower (but still fast overall), but in the relevant ones (many input events, bounded number of bins) it is not relevant.

File: tests/coords/transform_coords_test.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply tests/coords/transform_coords_test.py:1030]: Yes makes sense. You it be a property or a function returning a list?

[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply tests/coords/transform_coords_test.py:1030]: See update, I kept it as a property.

File: src/scipp/core/bins.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jl-wynen) [reply src/scipp/core/bins.py:33]: This doesn't show how the time relates to actually extracting values from the lookup table. But your reasoning makes sense. So I'm happy to leave it as is. We can revisit if we encounter problems.

File: tests/coords/transform_coords_test.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jl-wynen) [reply tests/coords/transform_coords_test.py:1030]: Looks good

**Review by jl-wynen — APPROVED** [review]
